<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot A Heatmap with ggplot2 — ggheatmap • Infercna</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot A Heatmap with ggplot2 — ggheatmap"><meta name="description" content="`ggheatmap()` create a heatmap ggplot2 object.
Uses `ggplot::geom_raster` to generate a plot of copy number aberration (CNA) values. Delineates chromosomes and chromosome arms, and provides options for clustering cells or groups of cells (e.g., cell types, samples, or subclones) prior to plotting."><meta property="og:description" content="`ggheatmap()` create a heatmap ggplot2 object.
Uses `ggplot::geom_raster` to generate a plot of copy number aberration (CNA) values. Delineates chromosomes and chromosome arms, and provides options for clustering cells or groups of cells (e.g., cell types, samples, or subclones) prior to plotting."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Infercna</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot A Heatmap with ggplot2</h1>

      <div class="d-none name"><code>cnaplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>`ggheatmap()` create a heatmap ggplot2 object.</p>
<p>Uses `ggplot::geom_raster` to generate a plot of copy number aberration (CNA) values. Delineates chromosomes and chromosome arms, and provides options for clustering cells or groups of cells (e.g., cell types, samples, or subclones) prior to plotting.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ggheatmap</span><span class="op">(</span></span>
<span>  <span class="va">dat</span>,</span>
<span>  x <span class="op">=</span> <span class="va">Var1</span>,</span>
<span>  y <span class="op">=</span> <span class="va">Var2</span>,</span>
<span>  fill <span class="op">=</span> <span class="va">value</span>,</span>
<span>  x.num <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  y.num <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  limits.find <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  limits.symmetric <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  na.value <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  raster.labels <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  raster.labels.col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  x.breaks <span class="op">=</span> <span class="fu">waiver</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  y.breaks <span class="op">=</span> <span class="fu">waiver</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  x.labels <span class="op">=</span> <span class="fu">waiver</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  y.labels <span class="op">=</span> <span class="fu">waiver</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  labels.col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  x.labels.angle <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  y.labels.angle <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  x.expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  y.expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  x.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  y.position <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>  x.title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  y.title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend.breaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend.labels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend.labels.col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  legend.width <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">0.3</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  legend.height <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">0.8</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  legend.margin <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="op">-</span><span class="fl">0.02</span>, unit <span class="op">=</span> <span class="st">"npc"</span><span class="op">)</span>,</span>
<span>  legend.title <span class="op">=</span> <span class="st">"Scaled Expression"</span>,</span>
<span>  legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  legend.justification <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>  title <span class="op">=</span> <span class="fu">waiver</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  subtitle <span class="op">=</span> <span class="fu">waiver</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  caption <span class="op">=</span> <span class="fu">waiver</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  text.size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  plot.margin <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span><span class="fl">0.15</span>, <span class="fl">0.15</span>, <span class="fl">0.15</span>, <span class="fl">0.15</span>, <span class="st">"cm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">cnaplot</span><span class="op">(</span></span>
<span>  <span class="va">cna</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  reorder <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reorder.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  groups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  interorder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dist.method <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  cluster.method <span class="op">=</span> <span class="st">"complete"</span>,</span>
<span>  hide <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"21"</span>, <span class="st">"Y"</span><span class="op">)</span>,</span>
<span>  limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend.title <span class="op">=</span> <span class="st">"Scaled CNA values"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Copy-number aberrations"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dat">dat<a class="anchor" aria-label="anchor" href="#arg-dat"></a></dt>
<dd><p>A <code>data.frame</code> or a <code>tibble</code> to plot ggplot.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>The column name of variable mapping to x axis.</p></dd>


<dt id="arg-y">y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>The column name of variable mapping to y axis.</p></dd>


<dt id="arg-fill">fill<a class="anchor" aria-label="anchor" href="#arg-fill"></a></dt>
<dd><p>The column name of variable mapping to heatmap fill.</p></dd>


<dt id="arg-x-num">x.num<a class="anchor" aria-label="anchor" href="#arg-x-num"></a></dt>
<dd><p>description</p></dd>


<dt id="arg-y-num">y.num<a class="anchor" aria-label="anchor" href="#arg-y-num"></a></dt>
<dd><p>description</p></dd>


<dt id="arg-limits">limits<a class="anchor" aria-label="anchor" href="#arg-limits"></a></dt>
<dd><p>for the color key; replaces out of bounds values with the nearest limit. Default: c(-1, 1)</p></dd>


<dt id="arg-limits-find">limits.find<a class="anchor" aria-label="anchor" href="#arg-limits-find"></a></dt>
<dd><p>If used, auto find the value limits to plot heatmap. Default: FALSE</p></dd>


<dt id="arg-limits-symmetric">limits.symmetric<a class="anchor" aria-label="anchor" href="#arg-limits-symmetric"></a></dt>
<dd><p>Used with <code>limits.find</code>. If used, set the value limits to be symmetrical. Default: FALSE</p></dd>


<dt id="arg-cols">cols<a class="anchor" aria-label="anchor" href="#arg-cols"></a></dt>
<dd><p>custom color palette (character vector), Default: NULL</p></dd>


<dt id="arg-na-value">na.value<a class="anchor" aria-label="anchor" href="#arg-na-value"></a></dt>
<dd><p>The colot of NA value.</p></dd>


<dt id="arg-raster-labels">raster.labels<a class="anchor" aria-label="anchor" href="#arg-raster-labels"></a></dt>
<dd><p>Whether label the value in heatmap. Default: FALSE</p></dd>


<dt id="arg-raster-labels-col">raster.labels.col<a class="anchor" aria-label="anchor" href="#arg-raster-labels-col"></a></dt>
<dd><p>The color of the label in heatmap</p></dd>


<dt id="arg-x-breaks-y-breaks">x.breaks, y.breaks<a class="anchor" aria-label="anchor" href="#arg-x-breaks-y-breaks"></a></dt>
<dd><p>same as <code>breaks</code> in <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></code> or <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></code></p></dd>


<dt id="arg-x-labels-y-labels">x.labels, y.labels<a class="anchor" aria-label="anchor" href="#arg-x-labels-y-labels"></a></dt>
<dd><p>same as <code>labels</code> in <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></code> or <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></code></p></dd>


<dt id="arg-labels-col">labels.col<a class="anchor" aria-label="anchor" href="#arg-labels-col"></a></dt>
<dd><p>The color of x axis and y axis label</p></dd>


<dt id="arg-x-labels-angle-y-labels-angle">x.labels.angle, y.labels.angle<a class="anchor" aria-label="anchor" href="#arg-x-labels-angle-y-labels-angle"></a></dt>
<dd><p>The Angle of x axis or y axis label (in \([0, 360]\))</p></dd>


<dt id="arg-x-expand-y-expand">x.expand, y.expand<a class="anchor" aria-label="anchor" href="#arg-x-expand-y-expand"></a></dt>
<dd><p>same as <code>expand</code> in <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></code> or <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></code></p></dd>


<dt id="arg-x-position-y-position">x.position, y.position<a class="anchor" aria-label="anchor" href="#arg-x-position-y-position"></a></dt>
<dd><p>same as <code>position</code> in <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></code> or <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></code></p></dd>


<dt id="arg-x-title-y-title">x.title, y.title<a class="anchor" aria-label="anchor" href="#arg-x-title-y-title"></a></dt>
<dd><p>The title of x axis or y axis</p></dd>


<dt id="arg-legend-breaks">legend.breaks<a class="anchor" aria-label="anchor" href="#arg-legend-breaks"></a></dt>
<dd><p>same as <code>breaks</code> in <code><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradientn</a></code></p></dd>


<dt id="arg-legend-labels">legend.labels<a class="anchor" aria-label="anchor" href="#arg-legend-labels"></a></dt>
<dd><p>same as <code>labels</code> in <code><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradientn</a></code></p></dd>


<dt id="arg-legend-labels-col">legend.labels.col<a class="anchor" aria-label="anchor" href="#arg-legend-labels-col"></a></dt>
<dd><p>The legend label color</p></dd>


<dt id="arg-legend-width">legend.width<a class="anchor" aria-label="anchor" href="#arg-legend-width"></a></dt>
<dd><p>same as <code>legend.width</code> in <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></code></p></dd>


<dt id="arg-legend-height">legend.height<a class="anchor" aria-label="anchor" href="#arg-legend-height"></a></dt>
<dd><p>same as <code>legend.height</code> in <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></code></p></dd>


<dt id="arg-legend-margin">legend.margin<a class="anchor" aria-label="anchor" href="#arg-legend-margin"></a></dt>
<dd><p>same as <code>legend.margin</code> in <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></code></p></dd>


<dt id="arg-legend-title">legend.title<a class="anchor" aria-label="anchor" href="#arg-legend-title"></a></dt>
<dd><p>legend title, Default: NULL</p></dd>


<dt id="arg-legend-position">legend.position<a class="anchor" aria-label="anchor" href="#arg-legend-position"></a></dt>
<dd><p>same as <code>legend.position</code> in <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></code></p></dd>


<dt id="arg-legend-justification">legend.justification<a class="anchor" aria-label="anchor" href="#arg-legend-justification"></a></dt>
<dd><p>same as <code>legend.justification</code> in <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></code></p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>plot title, Default: 'Copy-number aberrations'</p></dd>


<dt id="arg-subtitle">subtitle<a class="anchor" aria-label="anchor" href="#arg-subtitle"></a></dt>
<dd><p>same as <code>subtitle</code> in <code><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></code></p></dd>


<dt id="arg-caption">caption<a class="anchor" aria-label="anchor" href="#arg-caption"></a></dt>
<dd><p>same as <code>caption</code> in <code><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></code></p></dd>


<dt id="arg-text-size">text.size<a class="anchor" aria-label="anchor" href="#arg-text-size"></a></dt>
<dd><p>The base size of text in the plot</p></dd>


<dt id="arg-plot-margin">plot.margin<a class="anchor" aria-label="anchor" href="#arg-plot-margin"></a></dt>
<dd><p>same as <code>plot.margin</code> in <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></code></p></dd>


<dt id="arg-cna">cna<a class="anchor" aria-label="anchor" href="#arg-cna"></a></dt>
<dd><p>CNA matrix (genes by cells). It can be generated using <code><a href="Infercna.html">Infercna</a></code>.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>The genome annotation to order the genes.</p></dd>


<dt id="arg-reorder">reorder<a class="anchor" aria-label="anchor" href="#arg-reorder"></a></dt>
<dd><p>reorder cells using hierarchical clustering, Default: FALSE</p></dd>


<dt id="arg-reorder-by">reorder.by<a class="anchor" aria-label="anchor" href="#arg-reorder-by"></a></dt>
<dd><p>reorder cells by genes on select chromosomes or chromosome arms, e.g., c("7", "10", "1p", "19q"). Default: NULL</p></dd>


<dt id="arg-groups">groups<a class="anchor" aria-label="anchor" href="#arg-groups"></a></dt>
<dd><p>groups of cells to delineate and label on the plot (e.g., cell types, samples, or subclones). If &lt;reorder&gt; is TRUE, ordering will be performed within groups, Default: NULL</p></dd>


<dt id="arg-interorder">interorder<a class="anchor" aria-label="anchor" href="#arg-interorder"></a></dt>
<dd><p>reorder by hierarchical clustering betweeen groups, Default: TRUE</p></dd>


<dt id="arg-dist-method">dist.method<a class="anchor" aria-label="anchor" href="#arg-dist-method"></a></dt>
<dd><p>distance metric for reordering, Default: 'euclidean'</p></dd>


<dt id="arg-cluster-method">cluster.method<a class="anchor" aria-label="anchor" href="#arg-cluster-method"></a></dt>
<dd><p>linkage method for reordering, Default: 'complete'</p></dd>


<dt id="arg-hide">hide<a class="anchor" aria-label="anchor" href="#arg-hide"></a></dt>
<dd><p>hide x-axis labels for specific chromosomes (e.g., small chromosomes whose labels would overlap flanking labels), Default: c("21", "Y")</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other argument used in <code>ggheatmap</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>return a ggplot2 object</p>
<p>return a ggplot object</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jeason Zhao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

