# Package index

## All functions

- [`Infercna()`](https://moonerss.github.io/Infercna/reference/Infercna.md)
  : Infer Copy-Number Alterations From Single-Cell RNA-Seq Data

- [`Unlist()`](https://moonerss.github.io/Infercna/reference/Unlist.md)
  : Unlist, keeping original list or element names

- [`addGenome()`](https://moonerss.github.io/Infercna/reference/addGenome.md)
  : Add your own Genome

- [`availableGenomes()`](https://moonerss.github.io/Infercna/reference/availableGenomes.md)
  : List of available genomes

- [`bt771`](https://moonerss.github.io/Infercna/reference/bt771.md) :
  scRNA-seq data from a patient with Glioblastoma

- [`clip()`](https://moonerss.github.io/Infercna/reference/clip.md) :
  Squish matrix values into range

- [`cnaCor()`](https://moonerss.github.io/Infercna/reference/cnaCor.md)
  : Cell - Tumour CNA Correlations

- [`cnaHotspotCells()`](https://moonerss.github.io/Infercna/reference/cnaHotspotCells.md)
  : Find the Cells with the highest CNA signal

- [`cnaHotspotGenes()`](https://moonerss.github.io/Infercna/reference/cnaHotspotGenes.md)
  : Find the Genes with the highest CNA signal

- [`cnaScatterPlot()`](https://moonerss.github.io/Infercna/reference/cnaScatterPlot.md)
  : Visualise Malignant and Non-Malignant Subsets

- [`cnaSignal()`](https://moonerss.github.io/Infercna/reference/cnaSignal.md)
  : Calculate the Means of Squared CNA Values

- [`ggheatmap()`](https://moonerss.github.io/Infercna/reference/cnaplot.md)
  [`cnaplot()`](https://moonerss.github.io/Infercna/reference/cnaplot.md)
  : Plot A Heatmap with ggplot2

- [`colcenter()`](https://moonerss.github.io/Infercna/reference/colcenter.md)
  : Center a matrix column-wise

- [`comply()`](https://moonerss.github.io/Infercna/reference/comply.md)
  :

  Apply a Function to All Pairs of Elements in `x` or Between `x` and
  `y`

- [`currentGenome()`](https://moonerss.github.io/Infercna/reference/currentGenome.md)
  : List the current genome name

- [`.mostExpressedGenes()`](https://moonerss.github.io/Infercna/reference/dot-mostExpressedGenes.md)
  : Get most expression gene

- [`expandToClones()`](https://moonerss.github.io/Infercna/reference/expandToClones.md)
  : expand clones

- [`fetchModes()`](https://moonerss.github.io/Infercna/reference/fetchModes.md)
  : Find small clone

- [`filterGenes()`](https://moonerss.github.io/Infercna/reference/filterGenes.md)
  : Filter Genes by their Genome Features

- [`findClones()`](https://moonerss.github.io/Infercna/reference/findClones.md)
  : Find Clones

- [`findMalignant()`](https://moonerss.github.io/Infercna/reference/findMalignant.md)
  : Find Malignant Subset of Cells

- [`fitBimodal()`](https://moonerss.github.io/Infercna/reference/fitBimodal.md)
  : Fit a Bimodal Gaussian Distribution

- [`flip()`](https://moonerss.github.io/Infercna/reference/flip.md) :
  Flip the (nested) elements of a character vector (or list) with its
  names

- [`genesOn()`](https://moonerss.github.io/Infercna/reference/genesOn.md)
  : Retrieve Genes by their Genome Features

- [`grouped_reorder()`](https://moonerss.github.io/Infercna/reference/grouped_reorder.md)
  : Grouped Reorder and Ungroup

- [`has_dim()`](https://moonerss.github.io/Infercna/reference/has_dim.md)
  : check dim

- [`hca()`](https://moonerss.github.io/Infercna/reference/hca.md)
  [`hca_cor()`](https://moonerss.github.io/Infercna/reference/hca.md)
  [`hca_dist()`](https://moonerss.github.io/Infercna/reference/hca.md)
  [`hca_tree()`](https://moonerss.github.io/Infercna/reference/hca.md)
  [`hca_order()`](https://moonerss.github.io/Infercna/reference/hca.md)
  [`hca_groups()`](https://moonerss.github.io/Infercna/reference/hca.md)
  [`hca_reorder()`](https://moonerss.github.io/Infercna/reference/hca.md)
  : Hierarchical clustering analysis

- [`hg19`](https://moonerss.github.io/Infercna/reference/hg19.md) : hg19
  reference genome gene position annotation.

- [`hg38`](https://moonerss.github.io/Infercna/reference/hg38.md) : hg38
  reference genome gene position annotation.

- [`is_cor()`](https://moonerss.github.io/Infercna/reference/is_cor.md)
  : Check cor matrix

- [`mgh125`](https://moonerss.github.io/Infercna/reference/mgh125.md) :
  scRNA-seq data from a patient with Glioblastoma

- [`mm10`](https://moonerss.github.io/Infercna/reference/mm10.md) : mm10
  reference genome gene position annotation.

- [`mm39`](https://moonerss.github.io/Infercna/reference/mm39.md) : mm39
  reference genome gene position annotation.

- [`orderGenes()`](https://moonerss.github.io/Infercna/reference/orderGenes.md)
  : Order Genes by their Genomic Positions

- [`refCells`](https://moonerss.github.io/Infercna/reference/refCells.md)
  : Normal Cell IDs in a cohort of 28 Glioblastoma samples.

- [`refCorrect()`](https://moonerss.github.io/Infercna/reference/refCorrect.md)
  : Convert Relative CNA Values To Absolute

- [`retrieveGenome()`](https://moonerss.github.io/Infercna/reference/retrieveGenome.md)
  : Retrieve genome data

- [`rowcenter()`](https://moonerss.github.io/Infercna/reference/rowcenter.md)
  : Center a matrix row-wise

- [`runMean()`](https://moonerss.github.io/Infercna/reference/runMean.md)
  : Rolling Means

- [`splitGenes()`](https://moonerss.github.io/Infercna/reference/splitGenes.md)
  : Split Genes By Chromosome (Arm)

- [`subsplit()`](https://moonerss.github.io/Infercna/reference/subsplit.md)
  : Split a vector, list, matrix or dataframe into into a list of
  matrices or nested list according to substring matches.

- [`substri()`](https://moonerss.github.io/Infercna/reference/substri.md)
  : Extract substrings between delimiters by position

- [`theme_cna()`](https://moonerss.github.io/Infercna/reference/theme_scalop.md)
  : FUNCTION_TITLE

- [`unique_sample_names()`](https://moonerss.github.io/Infercna/reference/unique_sample_names.md)
  : Get unique sample names

- [`logtpm()`](https://moonerss.github.io/Infercna/reference/unlogtpm.md)
  [`unlogtpm()`](https://moonerss.github.io/Infercna/reference/unlogtpm.md)
  : Convert TPM to logTPM

- [`useGenome()`](https://moonerss.github.io/Infercna/reference/useGenome.md)
  : Select a Genome for infercna to use
