<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Infer Copy-Number Alterations From Single-Cell RNA-Seq Data — Infercna • Infercna</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Infer Copy-Number Alterations From Single-Cell RNA-Seq Data — Infercna"><meta name="description" content="Infer Copy-Number Alterations From Single-Cell RNA-Seq Data"><meta property="og:description" content="Infer Copy-Number Alterations From Single-Cell RNA-Seq Data"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Infercna</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Infer Copy-Number Alterations From Single-Cell RNA-Seq Data</h1>

      <div class="d-none name"><code>Infercna.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Infer Copy-Number Alterations From Single-Cell RNA-Seq Data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Infercna</span><span class="op">(</span></span>
<span>  <span class="va">expr_mat</span>,</span>
<span>  refCells <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  window <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  top_n <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>  noise <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  center.method <span class="op">=</span> <span class="st">"median"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-expr-mat">expr_mat<a class="anchor" aria-label="anchor" href="#arg-expr-mat"></a></dt>
<dd><p>a normalized expression matrix of genes X cells containing scRNA-seq expression data. It can be TPM, RPKM, CPM etc.</p></dd>


<dt id="arg-refcells">refCells<a class="anchor" aria-label="anchor" href="#arg-refcells"></a></dt>
<dd><p>a list of two or more character vectors, each containg cell IDs of normal cell types (one cell type per list element). Since these cell types are presumed normal cells, their CNA values can be used to correct the remaining CNA values. Note that all cell IDs should correspond to column names in expr_mat. See Infercna::refCells (normal cells) for an example. Default: NULL</p></dd>


<dt id="arg-window">window<a class="anchor" aria-label="anchor" href="#arg-window"></a></dt>
<dd><p>the size of the window to use for the rolling mean. Units are in number of genes. Default: 100</p></dd>


<dt id="arg-range">range<a class="anchor" aria-label="anchor" href="#arg-range"></a></dt>
<dd><p>values in expr_mat above and below this range will be set to values in range. See Infercna::clip() for more details. Default: c(-3, 3)</p></dd>


<dt id="arg-top-n">top_n<a class="anchor" aria-label="anchor" href="#arg-top-n"></a></dt>
<dd><p>a numeric value indicating if only top genes should be used in the CNA calculation and how many. if n is a fraction between 0 and 1, the number of genes included will n * nrow(expr_mat). Else the number of genes included will be n. Default: 3000</p></dd>


<dt id="arg-noise">noise<a class="anchor" aria-label="anchor" href="#arg-noise"></a></dt>
<dd><p>a numeric value, which if given, increases the boundaries within which CNA values are considered 0. Increases by &lt;noise&gt; i.e. the bounds become Minimum - noise and Maximum + noise. Default: 0.1</p></dd>


<dt id="arg-center-method">center.method<a class="anchor" aria-label="anchor" href="#arg-center-method"></a></dt>
<dd><p>method by which to center the cells after calculating CNA values. One of 'median', 'mean', etc.... Default: 'median'</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>print progress messages. Default: FALSE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a matrix of genes X cells of inferred CNA values. Note that n = (window - 1)/2 genes will be lost from either extremity of the genome (ie. n genes lost at the start of chromosome 1 and at the end of chromosome Y, if the genome in question is H.sapiens.)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Correction with reference cells' &lt;refCells&gt; CNAs: the boundaries of reference CNA values are the boundaries for what should be considered a CNA of 0. Thus, if the boundary is -0.1 and 0.1, then a cell with CNA = -0.5 will be corrected to -0.4, a cell with CNA value of 1 will be corrected to 0.9 and a cell with CNA value of 0.05 will be corrected to 0.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">mgh125</span></span></span>
<span class="r-in"><span><span class="va">cna</span> <span class="op">&lt;-</span> <span class="fu">Infercna</span><span class="op">(</span>expr_mat <span class="op">=</span> <span class="va">m</span>, refCells <span class="op">=</span> <span class="va">refCells</span>, top_n <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jeason Zhao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

